vpr/src/route/sink_sampling.h

```diff
@@ -0,0 +1,155 @@
+#pragma once
+
+/** @file Sink downsamplers for parallel routing. 
+ *
+ * These are used to get a "minimal skeleton routing" from the main task. 
+ * Rest of the routing is delegated to child tasks. They will work with a
+ * strictly limited bounding box, so it's necessary that the initial routing
+ * provides enough hints while routing to as few sinks as possible to limit
+ * the serial bottleneck. */
+
+#include <cmath>
+#include <limits>
+#include <vector>
+#include "globals.h"
+#include "partition_tree.h"
+#include "route_common.h"
+#include "router_lookahead_sampling.h"
+
+/** Sink container for geometry operations */
+struct SinkPoint {
+    int x;
+    int y;
+    int isink;
+
+    bool operator==(const SinkPoint& rhs) const {
+        return x == rhs.x && y == rhs.y;
+    }
+    bool operator<(const SinkPoint& rhs) const {
+        if (x < rhs.x)
+            return true;
+        if (x > rhs.x)
+            return false;
+        if (y < rhs.y)
+            return true;
+        if (y > rhs.y)
+            return false;
+        return isink < rhs.isink;
+    }
+};
+
+/* Namespace the following to keep these generic fn names available */
+namespace sink_sampling {
+
+/* TODO: Replace this convex hull fn with something better
+ * it was cobbled together quickly, probably not the most efficient code */
+
+/** Cross product of v0v1 and v0p */
+constexpr int det(const SinkPoint& p, const SinkPoint& v0, const SinkPoint& v1) {
+    return (v1.x - v0.x) * (p.y - v0.y) - (v1.y - v0.y) * (p.x - v0.x);
+}
+
+/** Which side of [v0, v1] has p? +1 is right, -1 is left */
+constexpr int side_of(const SinkPoint& p, const SinkPoint& v0, const SinkPoint& v1) {
+    return det(p, v0, v1) > 0 ? 1 : -1;
+}
+
+/** Perpendicular distance of p to v0v1 assuming |v0v1| = 1
+ * (it's not, so only use to compare when v0 and v1 is the same for different p's) */
+inline int dist(const SinkPoint& p, const SinkPoint& v0, const SinkPoint& v1) {
+    return abs(det(p, v0, v1));
+}
+
+/** Helper for quickhull() */
+inline void find_hull(std::set<SinkPoint>& out, const std::vector<SinkPoint>& points, const SinkPoint& v0, const SinkPoint& v1, int side) {
+    int max_dist = 0;
+    const SinkPoint* max_p = nullptr;
+    for (auto& point : points) {
+        if (side_of(point, v0, v1) != side) {
+            continue;
+        }
+        int h = dist(point, v0, v1);
```

Comment:
h is the height of a triangle?  Best to call it height then.  Also comment the overall algorithm here.